src/uu/base32/src/base_common.rs

```diff
@@ -134,52 +141,615 @@ pub fn base_app(about: &'static str, usage: &str) -> Command {
+                    {
+                        let bytes_to_chunk = if bytes_to_steal > 0_usize {
+                            let (stolen_bytes, rest_of_read_buffer_filtered) =
+                                read_buffer_filtered.split_at(bytes_to_steal);
+
+                            leftover_buffer.extend(stolen_bytes);
+
+                            // After appending the stolen bytes to `leftover_buffer`, it should be the right size
+                            assert!(leftover_buffer.len() == decode_in_chunks_of_size);
+
+                            // Decode the old un-decoded data and the stolen bytes, and add the result to
+                            // `decoded_buffer`
+                            supports_fast_decode
+                                .decode_into_vec(&leftover_buffer, &mut decoded_buffer)?;
+
+                            // Reset `leftover_buffer`
+                            leftover_buffer.clear();
+
+                            rest_of_read_buffer_filtered
+                        } else {
+                            // Do not need to steal bytes from `read_buffer`
+                            read_buffer_filtered
+                        };
+
+                        let chunks_exact = bytes_to_chunk.chunks_exact(decode_in_chunks_of_size);
+
+                        let remainder = chunks_exact.remainder();
+
+                        for sl in chunks_exact {
+                            assert!(sl.len() == decode_in_chunks_of_size);
+
+                            supports_fast_decode.decode_into_vec(sl, &mut decoded_buffer)?;
+                        }
+
+                        leftover_buffer.extend(remainder);
+                    }
+
+                    // Write all data in `decoded_buffer` to stdout
+                    write_to_stdout(&mut decoded_buffer, &mut stdout_lock)?;
+                }
+                Err(er) => {
+                    if er.kind() == ErrorKind::Interrupted {
+                        // TODO
+                        // Retry reading?
+                    }
+
+                    return Err(USimpleError::new(1_i32, format!("read error: {er}")));
+                }
             }
-            Err(EncodeError::InvalidInput) => Err(USimpleError::new(1, "error: invalid input")),
-            Err(_) => Err(USimpleError::new(
-                1,
-                "error: invalid input (length must be multiple of 4 characters)",
-            )),
         }
+
+        // Cleanup
+        // `input` has finished producing data, so the data remaining in the buffers needs to be decoded and printed
+        {
+            // Decode all remaining encoded bytes, placing them in `decoded_buffer`
+            supports_fast_decode.decode_into_vec(&leftover_buffer, &mut decoded_buffer)?;
+
+            // Write all data in `decoded_buffer` to stdout
+            write_to_stdout(&mut decoded_buffer, &mut stdout_lock)?;
+        }
+
+        Ok(())
```

Comment:
That's a lot of new complex code in `base_common.rs`. Could you add a few unit-tests, just to sanity-check everything? I know that it (seems to) work just fine, and skimming the code I didn't spot any obvious flaws.

- I'm sure there were a few edge cases where you had to take extra precaution. Those are ideal things to test for, because we want to make sure that now only is the *current* code correct, but also gives us assurance that future code *remains* correct.
- Also, Just basic functionality. This makes it easier to spot any issues at the right level, not only during the integration test of the base64 program (for example).

Reply:
What do you think of using fuzz or property testing? There was one attempted optimization I had to roll back because of a bug that only manifested on large inputs. The problem is that, AFAIK, the project is not using fuzz or property testing anywhere else, so it might be a bit out of place here.